### PR TITLE
mass backport

### PR DESCRIPTION
**Pull request by maan2003**
https://gist.github.com/maan2003/fd8494c01711b23aabc9ff28ec107d85
shows how I resolved the conflicts in all commits

backport 
- https://github.com/fedimint/fedimint/pull/4760
- https://github.com/fedimint/fedimint/pull/4783
- https://github.com/fedimint/fedimint/pull/4766
- https://github.com/fedimint/fedimint/pull/4785
- https://github.com/fedimint/fedimint/pull/4768
- https://github.com/fedimint/fedimint/pull/4787
- https://github.com/fedimint/fedimint/pull/4854
- https://github.com/fedimint/fedimint/pull/4864
- https://github.com/fedimint/fedimint/pull/4858
- https://github.com/fedimint/fedimint/pull/4838
- https://github.com/fedimint/fedimint/pull/4809
- https://github.com/fedimint/fedimint/pull/4799
- https://github.com/fedimint/fedimint/pull/4804
- https://github.com/fedimint/fedimint/pull/4828

### CODE REVIEW

**Comment by maan2003** [comment]
oh it doesn't compile, will fix later

**Comment by maan2003** [comment]
https://github.com/fedimint/fedimint/compare/5debf75571bbb804fa7aa18f3f53130bf477d12b..feeb96683de28eb9eee4113cf422167e9d54c941 is only significant change

File: fedimint-core/src/util/mod.rs [hunk 1]
```diff
@@ -319,42 +323,42 @@ pub fn handle_version_hash_command(version_hash: &str) {
 ///   error of the closure is returned
 pub async fn retry<F, Fut, T>(
     op_name: impl Into<String>,
+    statergy: impl backon::BackoffBuilder,
     op_fn: F,
-    interval: Duration,
-    max_attempts: u32,
 ) -> Result<T, anyhow::Error>
 where
```
Comment (elsirion) [inline fedimint-core/src/util/mod.rs:329]: Breaking change

File: modules/fedimint-ln-client/src/lib.rs [hunk 2]
```diff
@@ -899,8 +889,8 @@ impl LightningClientModule {
     /// from the federation.
     ///
     /// See also [`Self::update_gateway_cache_continuously`].
-    pub async fn update_gateway_cache(&self, apply_meta: bool) -> anyhow::Result<()> {
-        let gateways = self.fetch_registered_gateways(apply_meta).await?;
+    pub async fn update_gateway_cache(&self) -> anyhow::Result<()> {
```
Comment (elsirion) [inline modules/fedimint-ln-client/src/lib.rs:892]: breaking

File: modules/fedimint-ln-client/src/lib.rs [hunk 3]
```diff
@@ -922,7 +912,7 @@ impl LightningClientModule {
     /// Continuously update the gateway cache whenever a gateway expires.
     ///
     /// Client integrators are expected to call this function in a spawned task.
-    pub async fn update_gateway_cache_continuously(&self, apply_meta: bool) -> ! {
+    pub async fn update_gateway_cache_continuously(&self) -> ! {
```
Comment (elsirion) [inline modules/fedimint-ln-client/src/lib.rs:915]: breaking

File: modules/fedimint-ln-client/src/db.rs [hunk 4]
```diff
@@ -73,26 +71,6 @@ impl_db_record!(
 
 impl_db_lookup!(key = PaymentResultKey, query_prefix = PaymentResultPrefix);
 
-#[derive(Debug, Encodable, Decodable, Serialize)]
-pub struct MetaOverridesKey;
```
Comment (elsirion) [inline modules/fedimint-ln-client/src/db.rs:77]: technically breaking I guess, but it's not returned anywhere afaict

**Review by elsirion — COMMENTED** [review]
@maan2003 what do you think about deprecating old methods, but introducing the changed ones as new ones on this branch? That way we can keep SemVer happy and just slightly nudge users into using the new ones.

**Comment by maan2003** [comment]
> @maan2003 what do you think about deprecating old methods, but introducing the changed ones as new ones on this branch? That way we can keep SemVer happy and just slightly nudge users into using the new ones.

I think we should not guarantee semver for fedimint rust api. 
Breaking semver sucks for dependencies that are deep in dependency clain or are very popular. 
fedimint-client is neither, so it will be fine.

**Comment by elsirion** [comment]
@TonyGiorgio @benthecarman are you ok with breaking SemVer here? You are realistically the only ones who might have a little extra work, but you get a lot of functionality that otherwise each integrator would have to build themselves.

**Comment by TonyGiorgio** [comment]
> @TonyGiorgio @benthecarman are you ok with breaking SemVer here? You are realistically the only ones who might have a little extra work, but you get a lot of functionality that otherwise each integrator would have to build themselves.

Just from a library API perspective? Yeah we don't mind at all when it comes to "breaking" api changes across minor or major releases.

**Comment by maan2003** [comment]
lets wait for https://github.com/fedimint/fedimint/pull/4858, it is also going to break api

**Comment by maan2003** [comment]
@m1sterc001guy 89cc797ae318a0291ad8700284ed92bf7aaad90a didn't backport cleanly

**Comment by maan2003** [comment]
https://gist.github.com/maan2003/fd8494c01711b23aabc9ff28ec107d85
shows how I resolved the conflicts in all commits

**Review by dpc — APPROVED** [review]

File: devimint/src/tests.rs [hunk 5]
```diff
@@ -268,15 +268,32 @@ pub async fn latency_tests(dev_fed: DevFed, r#type: LatencyTest) -> Result<()> {
             let mut fm_internal_pay = Vec::with_capacity(iterations);
             let sender = fed.new_joined_client("internal-swap-sender").await?;
             fed.pegin_client(10_000_000, &sender).await?;
+            // TODO(support:v0.2): 0.3 removed the active gateway concept and requires an
+            // explicit gateway or `force-internal`
+            let fedimint_cli_version = crate::util::FedimintCli::version_or_default().await;
+            let default_internal = VersionReq::parse(">=0.3.0")?.matches(&fedimint_cli_version);
```
Comment (maan2003) [inline devimint/src/tests.rs:274]: this is messed up

**Comment by maan2003** [comment]
had to bump version for backwards compat tests to still work

**Review by elsirion — APPROVED** [review]

**Pull request merged**